### PR TITLE
Generate types from OpenAPI

### PR DESCRIPTION
**Pull request by NotWoods**
By using the OpenAPI spec inside the codebase, there's the combined effect of improving typing and improving the API spec.

`openapi-typescript-codegen` is a code generator written in Javascript. OpenAPI provides their own version, but it requires Java to be installed. I can switch the Pr to use this library if preferred.

I updated the API yaml file to use clear names, descriptions, and de-duplicated types. These files can then be imported inside of "src". 

Currently the generated files are put in "src/api-types", but the location doesn't matter. It might be preferable to generate a root folder and then import the types in the server code too.

The generate script is called from "prepare", which means it will automatically be run when you run `npm install` locally.

### CODE REVIEW

File: src/app/posts.services.ts [hunk 1]
```diff
@@ -25,7 +25,9 @@ export class PostsService implements CanActivate {
     // auth
     auth_token = '4241b401-7236-493e-92b5-b72696b9d853';
     session_id = null;
-    httpOptions = null;
+    httpOptions: {
```
Comment (Tzahi12345) [inline src/app/posts.services.ts:28]: Why this change?

**Review by Tzahi12345 — COMMENTED** [review]
I prefer a JS-only solution, so we're good with sticking with `openapi-typescript-codegen`. Thanks for updating the YAML! The refs needed better names, I just stuck with the autogenerated ones.

`api-types` is fine being in `src`, I can't think of a better place to put it. It would probably be prudent to commit the file though, right? Otherwise it will add an extra step to the manual (from source) installation and dev setup process.

File: src/app/posts.services.ts (hunk 1 above)
Comment (NotWoods) [reply src/app/posts.services.ts:28]: Two parts:
- This field is set in the constructor, so the default isn't needed.
- Angular uses the type of `httpOptions.responseType` to decide on the `get`/`post` return type. When this is set to `any`, `get`/`post` defaults to ArrayBuffer instead of JSON.

**Comment by NotWoods** [comment]
Every used route now has a named schema, and I committed the api-types.

Looks like the next version of `openapi-typescript-codegen` will let you use unions (`type X = 'a' | 'b'`) instead of enums (`enum X { A = 'a', B = 'b' }`). When enums are used, there are a few places where we need to cast a string to the enum type (`as FileType`). Those can be removed once union types are available.

**Comment by Tzahi12345** [comment]
>Every used route now has a named schema, and I committed the api-types.

Awesome!

>Looks like the next version of openapi-typescript-codegen will let you use unions (type X = 'a' | 'b') instead of enums (enum X { A = 'a', B = 'b' }). When enums are used, there are a few places where we need to cast a string to the enum type (as FileType). Those can be removed once union types are available.

Okay, we'll play that by ear, I have no problem transitioning to unions since it's more concise than enums.

I'll test out this PR over the weekend and merge if nothing comes up. Regarding your other PR, I found one minor issue but otherwise it looks great! Very happy with these changes

**Comment by Tzahi12345** [comment]
Noticed a couple things while testing:

1. Currently the code won't compile as the `import type {...}` syntax is TypeScript 3.8 only, and right now we're actually on TS 3.7. I do want to upgrade to Angular 10 sometime soon and so that will bring us to TS 4.0 so we'll be good then. I tried to do an upgrade yesterday and there were some conflicts so it's something I'll work on keep you updated. Hopefully in the next few days I can do this migration.

2. In some cases ([see here](https://github.com/Tzahi12345/YoutubeDL-Material/pull/218/files#diff-a7eb8c84a068b9c9c23a3a988dc39cf1R163)) FileType is being imported on variable assignment. This isn't the worst thing in the world, but in all of these cases the two options are either 'video' or 'audio'. I do agree that on variable *declaration* we should be casting those as `FileType`, but doing so on *assignment* seems overly verbose to me. Is there any major reason that I'm missing for casting on assignment?

**Comment by NotWoods** [comment]
1. Plain imports can be used, I'll update the code.
2. The casts are needed because we're generating enums, not union types. When union types are available the casts can be removed, but without it `'audio'` does not automatically equal `FileType.AUDIO`. We could replace the strings with the equivalent enum (see below) but then we'd need to switch it around again when switching to union types.

```diff
-  this.type = (this.db_file['isAudio'] ? 'audio' : 'video') as FileType;
+  this.type = this.db_file['isAudio'] ? FileType.AUDIO : FileType.VIDEO;
```

**Comment by NotWoods** [comment]
Branch has been replaced and `import type` has been removed.

I made a non-typing change: `downloadFileFromServer`'s argument list was getting intense, so I switched it to use named arguments. This helped catch some false assumptions I made in the related OpenAPI schema.

**Comment by Tzahi12345** [comment]
Apologies for waiting so long to merge, one weekend turned into a year :P

Things are pretty outdated so I'm going through and updating the API, I think this is a great way to keep the API docs and the actual functionality aligned.

So it's merged with the current master but won't build as not all the refs are updated, that will be the hard part. Will take a few days but it'll be worth it!
